docs/source/generics.rst

```diff
@@ -566,6 +564,59 @@ functions are typically small enough that this is not a big
 problem. This is also the reason for the :py:func:`~typing.cast` call in the
 ``return`` statement in ``my_decorator()``. See :ref:`casts`.
 
+.. _decorator-factories:
+
+Decorator factories
+-------------------
+
+Functions that take arguments and return a decorator (also called second-order decorators), are
+similarly supported via generics:
+
+.. code-block:: python
+
+    from typing import Any, Callable, TypeVar
+
+    F = TypeVar('F', bound=Callable)
+
+    def route(url: str) -> Callable[[F], F]:
+        ...
+
+    @route(url='/')
+    def index(request: Any) -> str:
+        return 'Hello world'
+
+Sometimes the same decorator supports both bare calls and calls with arguments. This can be
+achieved by combining with :py:func:`@overload <typing.overload>`:
+
+.. code-block:: python
+
+    from typing import Callable, TypeVar, overload
+
+    F = TypeVar('F', bound=Callable)
```

Comment:
Similar to above.

Reply:
Like this?

Should I keep the Callable unchanged in the function definition?:
```python
    def atomic(__func: Callable = None, *, savepoint: bool = True):
        def decorator(func: Callable):
```

Reply:
I'd use `Callable[..., Any]` everywhere instead of bare `Callable`, since it's just a more explicit way of saying the same thing. This is similar how we discourage users from writing `list` instead of `List[Any]`.